File: packages/agent-protocols/src/local-connector/state.ts
// In-memory store the connector projects records into. The public
// LocalConnectorState is the durable snapshot a host persists and restores; the
// room, inbox, and draft entries are the working set the handlers mutate.

import {
  AgentStatus,
  RoomJoinRequestStatus,
  RoomResponse,
  ServerRecord,
} from "../discourse.js";
import { AgentId } from "../identity.js";
import { AgentProfile } from "../profile.js";

import {
  ActiveTurn,
  AgentProtocolsHost,
  HeldDraft,
  InboxItem,
  RoomMemberView,
  TimelineItem,
} from "./views.js";
import { RoomSendMessageInput, RoomSubmitEventInput } from "./inputs.js";

/** `[host, roomId]`. Local room state is keyed by both: ADP room IDs are only
 * recommended to be globally unique and one connector can span hosts. */
export type RoomKey = { host: string; roomId: string };

/** Latest verified head-advancing record and locally applied chain tip. */
export interface LocalRoomState {
  host: string;
  room: RoomResponse;
  headSeq: number;
  headHash?: string;
  syncedSeq: number;
  syncedHash?: string;
  subscribed: boolean;
  members: Map<AgentId, RoomMemberView>;
  timeline: TimelineItem[];
  records: ServerRecord[];
  readSeq: number;
  activeTurn?: ActiveTurn;
}

export type InboxEntryState =
  | { kind: "pending" }
  | { kind: "claimed" }
  | { kind: "deferred"; until: number }
  | { kind: "acknowledged" };

export interface InboxEntry {
  item: InboxItem;
  state: InboxEntryState;
}

export type HeldDraftRequest =
  | { kind: "message"; input: RoomSendMessageInput }
  | { kind: "event"; input: RoomSubmitEventInput };

export interface HeldDraftEntry {
  draft: HeldDraft;
  request: HeldDraftRequest;
}

/**
 * Durable working set the connector projects records into. `hosts` is operator
 * configuration; the room, inbox, and draft maps are the materialized state a
 * host persists and restores through {@link LocalConnector} `state`.
 */
export class LocalConnectorState {
  hosts = new Map<string, AgentProtocolsHost>();
  rooms = new Map<string, LocalRoomState>();
  profiles = new Map<AgentId, AgentProfile>();
  joinRequests = new Map<string, RoomJoinRequestStatus[]>();
  agentStatuses = new Map<string, Map<AgentId, AgentStatus>>();
  inbox = new Map<string, InboxEntry>();
  drafts = new Map<string, HeldDraftEntry>();
}

File: crates/agent-protocols/src/local_connector/state.rs
//! In-memory store the local connector projects records into. The public
//! [`LocalConnectorState`] is the durable snapshot a host persists and restores;
//! the room, inbox, and draft entries are the crate-internal working set that
//! the handlers and projection read and mutate.

use std::collections::BTreeMap;

use crate::discourse::{AgentStatus, RoomJoinRequestStatus, RoomResponse, ServerRecord};
use crate::identity::AgentId;
use crate::profile::AgentProfile;

use super::inputs::{RoomSendMessageInput, RoomSubmitEventInput};
use super::views::{ActiveTurn, AgentProtocolsHost, HeldDraft, InboxItem, RoomMemberView, TimelineItem};

/// Local room state is keyed by `(host, room_id)`: ADP room IDs are only
/// RECOMMENDED to be globally unique, and a connector can be configured with
/// multiple hosts.
pub type RoomKey = (String, String);

#[derive(Clone, Debug, Default)]
pub struct LocalConnectorState {
    pub hosts: BTreeMap<String, AgentProtocolsHost>,
    pub(crate) rooms: BTreeMap<RoomKey, LocalRoomState>,
    pub(crate) profiles: BTreeMap<AgentId, AgentProfile>,
    pub(crate) join_requests: BTreeMap<RoomKey, Vec<RoomJoinRequestStatus>>,
    pub(crate) agent_statuses: BTreeMap<RoomKey, BTreeMap<AgentId, AgentStatus>>,
    pub(crate) inbox: BTreeMap<String, InboxEntry>,
    pub(crate) drafts: BTreeMap<String, HeldDraftEntry>,
}

impl LocalConnectorState {
    pub fn new() -> Self {
        Self::default()
    }
}

#[derive(Clone, Debug)]
pub(crate) struct LocalRoomState {
    pub(crate) host: String,
    pub(crate) room: RoomResponse,
    pub(crate) head_seq: u64,
    pub(crate) head_hash: Option<String>,
    pub(crate) synced_seq: u64,
    pub(crate) synced_hash: Option<String>,
    pub(crate) subscribed: bool,
    pub(crate) members: BTreeMap<AgentId, RoomMemberView>,
    pub(crate) timeline: Vec<TimelineItem>,
    pub(crate) records: Vec<ServerRecord>,
    pub(crate) read_seq: u64,
    pub(crate) active_turn: Option<ActiveTurn>,
}

impl LocalRoomState {
    pub(crate) fn new(host: String, room: RoomResponse) -> Self {
        Self {
            host,
            room,
            head_seq: 0,
            head_hash: None,
            synced_seq: 0,
            synced_hash: None,
            subscribed: false,
            members: BTreeMap::new(),
            timeline: Vec::new(),
            records: Vec::new(),
            read_seq: 0,
            active_turn: None,
        }
    }

    pub(crate) fn unread_count(&self) -> usize {
        self.timeline
            .iter()
            .filter(|item| item.seq > self.read_seq)
            .count()
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) enum InboxEntryState {
    Pending,
    Claimed,
    Deferred(i64),
    Acknowledged,
}

#[derive(Clone, Debug)]
pub(crate) struct InboxEntry {
    pub(crate) item: InboxItem,
    pub(crate) state: InboxEntryState,
}

#[derive(Clone, Debug)]
pub(crate) struct HeldDraftEntry {
    pub(crate) draft: HeldDraft,
    pub(crate) request: HeldDraftRequest,
}

#[derive(Clone, Debug)]
pub(crate) enum HeldDraftRequest {
    Message(RoomSendMessageInput),
    Event(RoomSubmitEventInput),
}
